### ISSUE TITLE
When i use the same code and datasets on different device training, the results are different，why？

### ISSUE THREAD

**Issue by my-littledarling** [ultralytics/ultralytics#20277]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

![Image](https://github.com/user-attachments/assets/6f36cd2a-ed16-457f-afbe-2cffc140a7d8)

![Image](https://github.com/user-attachments/assets/5a0185e3-9ba9-433e-9f5c-da135e0ef173)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @my-littledarling, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response. An Ultralytics engineer will also assist you here soon! 😊

**Comment by Y-T-G** [comment]
What was your training command?

**Comment by my-littledarling** [comment]
from ultralytics import YOLO

def main():
    model = YOLO('yolov8n.pt')
    # 训练模型
    train_results = model.train(
        data="data/s0.yaml",  # 数据配置文件的路径
        batch=16,
        epochs=250,  # 训练的轮数
        imgsz=640,  # 训练图像大小
        device=0,  # 运行的设备，例如 device=0 或 device=0,1,2,3 或 device=cpu
    )

if __name__ == '__main__':
    main()
-----------------------------------------------------------------------------
s0.yaml：
train: D:\PODB\gray\train # train images (relative to 'path') 4 images
val: D:\PODB\gray\val # val images (relative to 'path') 4 images
test : D:\PODB\gray\test
# Classes for DOTA 1.0
names:
  0: EV driver
  1: Electric tricycle
  2: Non-C electric tricycle
  3: Traffic warning signs
  4: car
  5: pedestrian
  6: traffic light
  7: truck

**Comment by my-littledarling** [comment]
![Image](https://github.com/user-attachments/assets/c6658192-f34a-4310-b745-7b9477f81d83)

**Comment by Y-T-G** [comment]
Do you have albumentations installed in one?

**Comment by glenn-jocher** [comment]
The difference in training results despite using the same code and datasets is likely due to randomness factors between your devices. Training neural networks involves multiple sources of non-determinism including:

1. Random weight initialization
2. Data shuffling during training
3. Hardware-specific floating-point operations

To achieve more consistent results across devices, try setting a fixed seed and enabling deterministic mode in your training command:

```python
model.train(data="path/to/data.yaml", epochs=100, seed=42, deterministic=True)
```

Or in CLI:
```
yolo detect train data=path/to/data.yaml model=yolo11n.pt epochs=100 seed=42 deterministic=True
```

This should help minimize differences, though you may still see slight variations due to hardware-specific implementations. If you need absolutely identical results, you would need identical hardware and software environments.
